cdap-common/src/main/java/co/cask/cdap/common/service/RetryOnStartFailureService.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.common.service;
+
+import com.google.common.base.Supplier;
+import com.google.common.util.concurrent.AbstractService;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.Service;
+import com.google.common.util.concurrent.Uninterruptibles;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ *
+ */
+public class RetryOnStartFailureService extends AbstractService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RetryOnStartFailureService.class);
+
+  private final Thread startupThread;
+  private final String delegateServiceName;
+  private volatile Service currentDelegate;
+
+  public RetryOnStartFailureService(final Supplier<Service> delegate, final RetryStrategy retryStrategy) {
```

Comment:
Does it make sense to add a default retry strategy so the retryStrategy here becomes nullable?